LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceFactory.cs

```diff
@@ -73,13 +74,32 @@ private async Task<LoRaDevice> RegisterCoreAsync(IoTHubDeviceInfo deviceInfo, Ca
                 {
                     throw new LoRaProcessingException("Failed to initialize device twins.", LoRaProcessingErrorCode.DeviceInitializationFailed);
                 }
+
                 this.loRaDeviceCache.Register(loRaDevice);
+
                 return loRaDevice;
             }
             catch
             {
-                this.connectionManager.Release(loRaDevice);
-                loRaDevice.Dispose();
+                // release the loradevice client explicitly. If we were unable to register, or there was already
+                // a connection registered, we will leak this client.
+                await loRaDeviceClient.DisposeAsync();
+
+                if (this.logger.IsEnabled(LogLevel.Debug))
+                {
+                    try
+                    {
+                        // if the created client is registered, release it
+                        if (loRaDeviceClient != ((IIdentityProvider<ILoRaDeviceClient>)this.connectionManager.GetClient(loRaDevice)).Identity)
```

Comment:
@p-schuler should we use ReferenceEquals here?

Reply:
That's what will happen (a ref compare). nit, but we can make it explicit so ppl understand.

Reply:
Yes, it was a nit. Just asking if it was fine for you, since you wrote that part. Done.